pkg/agent/flowexporter/exporter/exporter.go

```diff
@@ -148,56 +146,138 @@ func NewFlowExporter(connStore *connections.ConntrackConnectionStore, records *f
-		exp.process.CloseConnToCollector()
-		exp.process = nil
-		return
-	}
-	klog.V(2).Infof("Successfully exported IPFIX flow records")
+}
+
+// getExpiryFromExpirePriorityQueue returns the shortest expire time duration
+// from expire priority queue.
+func (exp *flowExporter) getExpiryFromExpirePriorityQueue() time.Duration {
+	currTime := time.Now()
+	if exp.expirePriorityQueue.Len() > 0 {
+		// Get the minExpireTime of the top item in ExpirePriorityQueue.
+		expiryDuration := MinExpiryTime + exp.expirePriorityQueue.MinExpireTime(0).Sub(currTime)
+		if expiryDuration < 0 {
+			return MinExpiryTime
+		}
+		return expiryDuration
+	}
+	if exp.expirePriorityQueue.ActiveFlowTimeout < exp.expirePriorityQueue.IdleFlowTimeout {
+		return exp.expirePriorityQueue.ActiveFlowTimeout
+	}
+	return exp.expirePriorityQueue.IdleFlowTimeout
+}
+
+func (exp *flowExporter) sendFlowRecords() (time.Duration, error) {
+	exp.conntrackConnStore.AcquireConnStoreLock()
+	exp.denyConnStore.AcquireConnStoreLock()
+
+	expiredConns := make([]flowexporter.Connection, 0)
+	currTime := time.Now()
+	for {
+		// If the queue is empty or top item is not timeout, then we do not have to
+		// check the following items.
+		topItem := exp.expirePriorityQueue.Peek()
+		if topItem == nil || (topItem.ActiveExpireTime.After(currTime) && topItem.IdleExpireTime.After(currTime)) {
+			break
+		}
+		pqItem := heap.Pop(exp.expirePriorityQueue).(*flowexporter.ItemToExpire)
+		expiredConns = append(expiredConns, *pqItem.Conn)
+		if !pqItem.Conn.IsDenyConn && flowexporter.IsConnectionDying(pqItem.Conn) {
+			// If the conntrack connection is in dying state or connection
+			// is not in conntrack table, we set the ReadyToDelete flag to
+			// true to do the deletion later.
+			pqItem.Conn.ReadyToDelete = true
+		}
+		if pqItem.Conn.IsDenyConn && pqItem.IdleExpireTime.Before(currTime) {
+			// If the deny connection is idle time out, we set the ReadyToDelete
+			// flag to true to do the deletion later.
+			pqItem.Conn.ReadyToDelete = true
+		}
+		if !pqItem.Conn.IsDenyConn && pqItem.IdleExpireTime.Before(currTime) {
+			pqItem.Conn.IsActive = false
+		}
+		if pqItem.Conn.IsDenyConn && pqItem.Conn.OriginalPackets <= pqItem.Conn.PrevPackets {
+			pqItem.Conn.IsActive = false
+		}
+		if pqItem.Conn.ReadyToDelete || !pqItem.Conn.IsActive {
+			exp.removeConnFromMap(pqItem.Conn)
+		} else {
+			// For active connections, we update their "prev" stats fields,
+			// reset active expire time and push back into the PQ.
+			exp.updateConnStats(pqItem.Conn)
+			exp.resetActiveExpireTime(exp.expirePriorityQueue, pqItem, currTime)
+		}
+	}
+	nextExpireTime := exp.getExpiryFromExpirePriorityQueue()
+	exp.conntrackConnStore.ReleaseConnStoreLock()
+	exp.denyConnStore.ReleaseConnStoreLock()
+
+	for _, conn := range expiredConns {
+		if err := exp.exportConn(conn); err != nil {
+			klog.Errorf("Error when sending expired flow record: %v", err)
+			return nextExpireTime, err
+		}
+		conn.LastExportTime = currTime
```

Comment:
Because connections in the `expiredConns` are copies of connection instead of pointer, I'm worried whether this assignment will take effect.

Reply:
Thanks for catching that! Moved to lines above.